Marlin/src/lcd/dogm/marlinui_DOGM.cpp

```diff
@@ -447,11 +447,35 @@ void MarlinUI::clear_lcd() { } // Automatically cleared by Picture Loop
   }
 
   // Draw a generic menu item
-  void MenuItemBase::_draw(const bool sel, const uint8_t row, FSTR_P const ftpl, const char, const char post_char) {
+  void MenuItemBase::_draw(const bool sel, const uint8_t row, FSTR_P const ftpl, const char, const char post_char, const uint8_t style, const char *vstr, const uint8_t minFstr) {
     if (mark_as_selected(row, sel)) {
-      pixel_len_t n = lcd_put_u8str(ftpl, itemIndex, itemStringC, itemStringF, LCD_WIDTH - 1) * (MENU_FONT_WIDTH);
-      while (n > MENU_FONT_WIDTH) n -= lcd_put_u8str(F(" "));
-      lcd_put_lchar(LCD_PIXEL_WIDTH - (MENU_FONT_WIDTH), row_y2, post_char);
+      const pixel_len_t rlen = vstr ? utf8_strlen(vstr) + 1 : 0;
+      const uint8_t post_char_len = post_char != ' ' ? 1 : 0;
+      pixel_len_t n = _MAX(LCD_WIDTH - rlen - post_char_len, 0);
+
+      const bool full = bool(style & SS_FULL), center = bool(style & SS_CENTER);
+
+      if (!full || !vstr) {
+
+        const uint8_t totalLen = rlen + utf8_strlen(ftpl);
+        uint8_t padLeft = center ? _MAX(0, (LCD_WIDTH - post_char_len - totalLen) / 2) : 0;
+        n = LCD_WIDTH - post_char_len - padLeft;
+        while (padLeft > 0) padLeft -= lcd_put_u8str(F(" "))/MENU_FONT_WIDTH;
+        n -= lcd_put_u8str(ftpl, itemIndex, itemStringC, itemStringF, n);
+        n *= MENU_FONT_WIDTH;
+        if (vstr) lcd_put_u8str_max(vstr, n);
+        while (n > 0) n -= lcd_put_u8str(F(" "));
+
+      }
+      else {
+
+        n -= lcd_put_u8str(ftpl, itemIndex, itemStringC, itemStringF, n);
+        n *= MENU_FONT_WIDTH;
+        while (n > 0) n -= lcd_put_u8str(F(" "));
```

Comment:
Multiplying by `MENU_FONT_WIDTH` looks odd here.
If `lcd_put_u8str(` returns the number of pixels, then multiplication should not be needed.
However, if `lcd_put_u8str(` returns the number of chars, then the line 474 looks odd as it

Reply:
Did you mean the following?

```c
        n *= MENU_FONT_WIDTH;
        n -= lcd_put_u8str(ftpl, itemIndex, itemStringC, itemStringF, n);
        while (n > 0) n -= lcd_put_u8str(F(" "));
```

Reply:
The oddity is not the "out of place" multiplication... if I remember correctly the oddity is that lcd_put_u8str is not a single method - it is an overloaded method that returns different based on which overload it is. The first one (with string substitution) returns number of chars and the second one (with just single arg) returns pixels.

So my code is correct and if we want to fix the semantics, it would mean digging into the mess of overloads with different return specs 🙂.